tests/integration_test.rs

```diff
@@ -0,0 +1,439 @@
+
+    // Compare the return string
+    let listener_string = upclient
+        .register_rpc_listener(uuri.clone(), Box::new(|_| {}))
+        .await
+        .unwrap();
+    assert_eq!(listener_string, "upl/0100162e04d20100_0");
+
+    // Able to ungister
+    let result = upclient
+        .unregister_rpc_listener(uuri.clone(), &listener_string)
+        .await
+        .unwrap();
+    assert_eq!(result, ());
+
+    // Unable to ungister
+    let result = upclient
+        .unregister_rpc_listener(uuri.clone(), &listener_string)
+        .await;
+    assert_eq!(
+        result,
+        Err(UStatus::fail_with_code(
+            UCode::INVALID_ARGUMENT,
+            "RPC request listener doesn't exist"
+        ))
+    )
+}
+
+#[async_std::test]
+async fn test_utransport_special_uuri_register_and_unregister() {
+    let upclient = UPClientZenoh::new(Config::default()).await.unwrap();
+    let uuri = create_special_uuri();
+
+    // Compare the return string
+    let listener_string = upclient
+        .register_listener(uuri.clone(), Box::new(|_| {}))
+        .await
+        .unwrap();
+    assert_eq!(
+        listener_string,
+        "upr/060102030a0b0c/**_0&upr/060102030a0b0c/**_1&upr/060102030a0b0c/**_2"
+    );
+
+    // Able to ungister
+    let result = upclient
+        .unregister_listener(uuri.clone(), &listener_string)
+        .await
+        .unwrap();
+    assert_eq!(result, ());
+
+    // Unable to ungister
+    let result = upclient
+        .unregister_listener(uuri.clone(), &listener_string)
+        .await;
+    assert_eq!(
+        result,
+        Err(UStatus::fail_with_code(
+            UCode::INVALID_ARGUMENT,
+            "RPC response callback doesn't exist"
+        ))
+    )
+}
+
+#[async_std::test]
+async fn test_publish_and_subscribe() {
+    let target_data = String::from("Hello World!");
+    let upclient = UPClientZenoh::new(Config::default()).await.unwrap();
+    let uuri = create_utransport_uuri();
+
+    // Register the listener
+    let uuri_cloned = uuri.clone();
+    let data_cloned = target_data.clone();
+    let listener = move |result: Result<UMessage, UStatus>| match result {
+        Ok(msg) => {
+            if let Data::Value(v) = msg.payload.unwrap().data.unwrap() {
+                let value = v.into_iter().map(|c| c as char).collect::<String>();
+                assert_eq!(msg.attributes.unwrap().sink.unwrap(), uuri_cloned);
+                assert_eq!(value, data_cloned);
+            } else {
+                panic!("The message should be Data::Value type.");
+            }
+        }
+        Err(ustatus) => panic!("Internal Error: {:?}", ustatus),
+    };
+    let listener_string = upclient
+        .register_listener(uuri.clone(), Box::new(listener))
+        .await
+        .unwrap();
+
+    // Create uattributes
+    let mut attributes = UAttributesBuilder::publish(UPriority::UPRIORITY_CS4).build();
+    attributes.sink = Some(uuri.clone()).into();
```

Comment:
Making a link to show there's a bit of a gap here -- the sink can be left blank for a publish style message, and if it were, currently a listener for this UUri could not find it, since the [sink is being used up above here](https://github.com/eclipse-uprotocol/up-client-zenoh-rust/pull/2/files#r1508078000) for publish-style message, which it should be the source.

I think you'd need to swap sink for source up above there @evshary

The [issue @sophokles73 opened up over here](https://github.com/eclipse-uprotocol/up-spec/issues/73) made me take a closer look.